lib/xnvme_be_macos_driverkit_async.c

```diff
@@ -0,0 +1,196 @@
+// SPDX-FileCopyrightText: Samsung Electronics Co., Ltd
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <libxnvme.h>
+#include <xnvme_be.h>
+#include <xnvme_be_nosys.h>
+#ifdef XNVME_BE_MACOS_ENABLED
+#include <errno.h>
+#include <xnvme_be_macos_driverkit.h>
+#include <mach/mach_error.h>
+
+#include <xnvme_dev.h>
+#include <xnvme_queue.h>
+
+struct xnvme_queue_macos_driverkit {
+	struct xnvme_queue_base base;
+	RingQueue *sq_queue;
+	RingQueue *cq_queue;
+
+	uint id;
+
+	uint8_t _rsvd[208];
+};
+XNVME_STATIC_ASSERT(sizeof(struct xnvme_queue_macos_driverkit) == XNVME_BE_QUEUE_STATE_NBYTES,
+		    "Incorrect size")
+
+int
+xnvme_be_macos_driverkit_queue_init(struct xnvme_queue *_queue, int XNVME_UNUSED(opts))
+{
+	struct xnvme_queue_macos_driverkit *q = (struct xnvme_queue_macos_driverkit *)_queue;
+	struct xnvme_be_macos_driverkit_state *state =
+		(struct xnvme_be_macos_driverkit_state *)_queue->base.dev->be.state;
+
+	int qpid;
+	qpid = _xnvme_be_driverkit_create_ioqpair(state, q->base.capacity, 0);
+	if (qpid < 0) {
+		XNVME_DEBUG("FAILED: nvme_create_ioqpair(); err: %d", qpid);
+		return qpid;
+	}
+	q->id = qpid;
+
+	mach_vm_size_t size = 0;
+	mach_vm_address_t addr;
+	kern_return_t res;
+	res = IOConnectMapMemory64(state->device_connection, 1 << 16 | qpid, mach_task_self(),
+				   &addr, &size, kIOMapAnywhere);
+	XNVME_DEBUG("IOConnectMapMemory64 returned: %x, %s", res, mach_error_string(res));
+	assert(res == kIOReturnSuccess);
+	q->sq_queue = (RingQueue *)addr;
+	assert(q->sq_queue->buffer_size == size);
+
+	size = 0;
+	res = IOConnectMapMemory64(state->device_connection, qpid, mach_task_self(), &addr, &size,
+				   kIOMapAnywhere);
+	XNVME_DEBUG("IOConnectMapMemory64 returned: %x, %s", res, mach_error_string(res));
+	assert(res == kIOReturnSuccess);
+	q->cq_queue = (RingQueue *)addr;
+	assert(q->cq_queue->buffer_size == size);
+
+	return 0;
+}
+
+int
+xnvme_be_macos_driverkit_queue_term(struct xnvme_queue *_queue)
+{
+	struct xnvme_queue_macos_driverkit *q = (struct xnvme_queue_macos_driverkit *)_queue;
+	struct xnvme_be_macos_driverkit_state *state =
+		(struct xnvme_be_macos_driverkit_state *)q->base.dev->be.state;
+	_xnvme_be_driverkit_delete_ioqpair(state, q->id);
+
+	q->cq_queue = NULL;
+	q->sq_queue = NULL;
+	return 0;
+}
+
+int
+xnvme_be_macos_driverkit_queue_poke(struct xnvme_queue *queue, uint32_t XNVME_UNUSED(max))
+{
+	int queue_ret;
+	struct xnvme_queue_macos_driverkit *q = (struct xnvme_queue_macos_driverkit *)queue;
+	struct xnvme_be_macos_driverkit_state *state =
+		(struct xnvme_be_macos_driverkit_state *)queue->base.dev->be.state;
+	uint64_t qid[2] = {q->id, q->id}; // Submission queue, completion queue
+
+	kern_return_t ret =
+		IOConnectCallScalarMethod(state->device_connection, NVME_POKE, qid, 2, NULL, 0);
+	if (ret != kIOReturnSuccess) {
```

Comment:
See notes on declaration / initialization order and error-handling.

Reply:
Done.